File: Apps/Android/app/src/main/java/com/wikirace/bot/WikiRaceClient.kt
package com.wikirace.bot

import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import org.json.JSONArray

data class SessionInfo(val id: String, val secretToken: String)
data class PageRef(val lang: String, val title: String, val pageid: Long? = null)

class WikiRaceClient {
    private val BASE = "https://wiki-race.com"
    private val JSON = "application/json".toMediaType()

    private val http = OkHttpClient.Builder()
        .addInterceptor { chain ->
            chain.proceed(chain.request().newBuilder()
                .header("User-Agent", "Mozilla/5.0 (Linux; Android 14) Chrome/120.0.0.0 Mobile Safari/537.36")
                .build())
        }
        .build()

    data class JoinResult(val gameId: String, val session: SessionInfo, val playerName: String)

    fun joinGame(gameId: String, playerName: String): JoinResult {
        val body = JSONObject().put("gameId", gameId).put("playerName", playerName)
        val resp = post("/api/game/join", body)
        return JoinResult(
            gameId     = resp.getString("gameId"),
            session    = parseSession(resp),
            playerName = resp.optString("playerName", playerName),
        )
    }

    fun createGame(playerName: String): JoinResult {
        val body = JSONObject().put("playerName", playerName)
        val resp = post("/api/game", body)
        return JoinResult(
            gameId     = resp.getString("gameId"),
            session    = parseSession(resp),
            playerName = playerName,
        )
    }

    fun submitPath(gameId: String, session: SessionInfo, path: List<PageRef>) {
        val pathArr = JSONArray().also { arr ->
            path.forEach { p ->
                arr.put(JSONObject().put("title", p.title).apply {
                    if (p.pageid != null) put("pageid", p.pageid) else put("pageid", JSONObject.NULL)
                })
            }
        }
        post("/api/game/location", JSONObject()
            .put("gameId", gameId)
            .put("session", sessionJson(session))
            .put("path", pathArr))
    }

    fun surrender(gameId: String, session: SessionInfo) =
        post("/api/game/surrender", JSONObject().put("gameId", gameId).put("session", sessionJson(session)))

    fun continueGame(gameId: String, session: SessionInfo) =
        post("/api/game/continue", JSONObject().put("gameId", gameId).put("session", sessionJson(session)))

    fun pusherAuth(gameId: String, session: SessionInfo, socketId: String, channelName: String): JSONObject {
        val formBody = FormBody.Builder()
            .add("socket_id", socketId)
            .add("channel_name", channelName)
            .add("sessionId", session.id)
            .add("secretToken", session.secretToken)
            .add("gameId", gameId)
            .build()
        val req = Request.Builder().url("$BASE/api/game/pusher/auth").post(formBody).build()
        http.newCall(req).execute().use { resp ->
            val bodyStr = resp.body?.string() ?: "{}"
            if (!resp.isSuccessful) throw Exception("Pusher auth failed ${resp.code}: $bodyStr")
            return JSONObject(bodyStr)
        }
    }

    fun fetchSnapshot(gameId: String, session: SessionInfo): JSONObject {
        val url = "$BASE/game?gameId=$gameId&sessionId=${session.id}&secretToken=${session.secretToken}"
        val req = Request.Builder().url(url).header("Accept", "text/html").build()
        http.newCall(req).execute().use { resp ->
            val html = resp.body?.string() ?: ""
            val m = Regex("""<script id="__NEXT_DATA__" type="application/json">(.*?)</script>""",
                RegexOption.DOT_MATCHES_ALL).find(html)
                ?: throw Exception("__NEXT_DATA__ not found")
            val root = JSONObject(m.groupValues[1])
            return root.getJSONObject("props").getJSONObject("pageProps")
        }
    }

    fun waitForState(gameId: String, session: SessionInfo, state: String, timeoutMs: Long = 600_000): JSONObject {
        val deadline = System.currentTimeMillis() + timeoutMs
        var last: String? = null
        while (true) {
            val snap = fetchSnapshot(gameId, session)
            val cur = snap.optJSONObject("game")?.optString("state")
            if (cur != last) { log("[lobby] state=$cur"); last = cur }
            if (cur == state) return snap
            if (System.currentTimeMillis() > deadline) throw Exception("Timeout waiting for $state")
            Thread.sleep(2000)
        }
    }

    // ── helpers ──────────────────────────────────────────────────────────

    private fun post(path: String, body: JSONObject): JSONObject {
        val req = Request.Builder()
            .url("$BASE$path")
            .post(body.toString().toRequestBody(JSON))
            .header("Content-Type", "application/json")
            .build()
        http.newCall(req).execute().use { resp ->
            val str = resp.body?.string() ?: "{}"
            if (!resp.isSuccessful) throw Exception("POST $path → ${resp.code}: $str")
            return JSONObject(str)
        }
    }

    private fun parseSession(resp: JSONObject): SessionInfo {
        val s = resp.getJSONObject("session")
        return SessionInfo(s.getString("id"), s.getString("secretToken"))
    }

    private fun sessionJson(s: SessionInfo) =
        JSONObject().put("id", s.id).put("secretToken", s.secretToken)

    // Overridable for UI logging
    open fun log(msg: String) = println(msg)
}

File: Apps/Windows/WikiRaceClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace WikiRaceBot;

class WikiRaceClient : IDisposable
{
    private readonly HttpClient _http;
    private const string BaseUrl = "https://wiki-race.com";

    public WikiRaceClient()
    {
        _http = new HttpClient();
        _http.DefaultRequestHeaders.Add("User-Agent",
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0 Safari/537.36");
        _http.DefaultRequestHeaders.Add("Accept", "application/json");
        _http.Timeout = TimeSpan.FromSeconds(20);
    }

    public async Task<(string GameId, SessionInfo Session, string PlayerName)> JoinGameAsync(
        string gameId, string playerName)
    {
        var resp = await PostAsync("/api/game/join", new { gameId, playerName });
        var session = ParseSession(resp);
        var actualName = resp["playerName"]?.GetValue<string>() ?? playerName;
        return (resp["gameId"]!.GetValue<string>(), session, actualName);
    }

    public async Task<(string GameId, SessionInfo Session)> CreateGameAsync(string playerName)
    {
        var resp = await PostAsync("/api/game", new { playerName });
        return (resp["gameId"]!.GetValue<string>(), ParseSession(resp));
    }

    public async Task StartGameAsync(string gameId, SessionInfo session, GameSettings settings)
    {
        await PostAsync("/api/game/start", new
        {
            gameId,
            session = session.ToPayload(),
            settings = new
            {
                language = settings.Language,
                start    = new { title = settings.Start.Title,       pageid = settings.Start.PageId },
                destination = new { title = settings.Destination.Title, pageid = settings.Destination.PageId },
            }
        });
    }

    public async Task SubmitPathAsync(string gameId, SessionInfo session, IEnumerable<PageRef> path)
    {
        await PostAsync("/api/game/location", new
        {
            gameId,
            session = session.ToPayload(),
            path    = path.Select(p => p.ToPathEntry()).ToArray(),
        });
    }

    public async Task SurrenderAsync(string gameId, SessionInfo session) =>
        await PostAsync("/api/game/surrender", new { gameId, session = session.ToPayload() });

    public async Task ContinueAsync(string gameId, SessionInfo session) =>
        await PostAsync("/api/game/continue", new { gameId, session = session.ToPayload() });

    // Pusher auth — form-encoded POST body
    public async Task<JsonObject> PusherAuthAsync(
        string gameId, SessionInfo session, string socketId, string channelName)
    {
        var form = new FormUrlEncodedContent(new Dictionary<string, string>
        {
            ["socket_id"]    = socketId,
            ["channel_name"] = channelName,
            ["sessionId"]    = session.Id,
            ["secretToken"]  = session.SecretToken,
            ["gameId"]       = gameId,
        });
        var r = await _http.PostAsync($"{BaseUrl}/api/game/pusher/auth", form);
        r.EnsureSuccessStatusCode();
        var json = await r.Content.ReadAsStringAsync();
        return JsonNode.Parse(json)!.AsObject();
    }

    // Poll game state from __NEXT_DATA__
    public async Task<JsonObject> FetchSnapshotAsync(string gameId, SessionInfo session)
    {
        var url = $"{BaseUrl}/game?gameId={gameId}&sessionId={session.Id}&secretToken={session.SecretToken}";
        var html = await _http.GetStringAsync(url);
        var m = System.Text.RegularExpressions.Regex.Match(
            html, @"<script id=""__NEXT_DATA__"" type=""application/json"">(.*?)</script>",
            System.Text.RegularExpressions.RegexOptions.Singleline);
        if (!m.Success) throw new Exception("__NEXT_DATA__ not found");
        var root = JsonNode.Parse(m.Groups[1].Value)!;
        return root["props"]!["pageProps"]!.AsObject();
    }

    public async Task<JsonObject> WaitForStateAsync(
        string gameId, SessionInfo session, string desiredState,
        int timeoutSeconds = 600, int pollMs = 2000)
    {
        var deadline = DateTime.UtcNow.AddSeconds(timeoutSeconds);
        string? last = null;
        while (true)
        {
            var snap = await FetchSnapshotAsync(gameId, session);
            var state = snap["game"]?["state"]?.GetValue<string>();
            if (state != last) { Console.WriteLine($"[lobby] state={state}"); last = state; }
            if (state == desiredState) return snap;
            if (DateTime.UtcNow > deadline) throw new TimeoutException($"Timed out waiting for {desiredState}");
            await Task.Delay(pollMs);
        }
    }

    // ── helpers ──────────────────────────────────────────────────────

    private async Task<JsonObject> PostAsync(string path, object payload)
    {
        var resp = await _http.PostAsJsonAsync($"{BaseUrl}{path}", payload);
        var body = await resp.Content.ReadAsStringAsync();
        if (!resp.IsSuccessStatusCode)
            throw new Exception($"POST {path} → HTTP {(int)resp.StatusCode}: {body}");
        return JsonNode.Parse(body)!.AsObject();
    }

    private static SessionInfo ParseSession(JsonObject resp)
    {
        var s = resp["session"]!.AsObject();
        return new SessionInfo(s["id"]!.GetValue<string>(), s["secretToken"]!.GetValue<string>());
    }

    public void Dispose() => _http.Dispose();
}
